pkg/gadgets/socket-collector/gadget.go

```diff
@@ -72,71 +75,121 @@ func (t *Trace) Operation(trace *gadgetv1alpha1.Trace,
+		gadgets.SetStatusError(trace, fmt.Sprintf("This gadget only accepts operations on traces in the %s namespace",
+			gadgets.TRACE_DEFAULT_NAMESPACE))
 		return
 	}
 
 	switch operation {
 	case "start":
-		var pid uint32 = 0
-
 		if trace.Spec.Filter == nil {
 			gadgets.SetStatusError(trace, "Missing filter")
 			return
 		}
-		if trace.Spec.Filter.Namespace == "" {
-			gadgets.SetStatusError(trace, "Invalid filter: missing namespace")
-			return
+		if trace.Spec.Filter.ContainerName != "" {
+			log.Warningf("Gadget %s: Container name filter is not applicable in this gadget, ignoring it!",
+				trace.Spec.Gadget)
 		}
-		if trace.Spec.Filter.Podname == "" {
-			gadgets.SetStatusError(trace, "TODO: Filtering only by namespace is not currently supported")
+
+		selector := gadgets.ContainerSelectorFromContainerFilter(trace.Spec.Filter)
+		filteredContainers := t.resolver.GetContainersBySelector(selector)
+		if len(filteredContainers) == 0 {
+			gadgets.SetStatusError(trace, fmt.Sprintf("Any container matched the requested filter: %+v",
+				*trace.Spec.Filter))
 			return
 		}
-		if len(trace.Spec.Filter.Labels) != 0 {
-			gadgets.SetStatusError(trace, "TODO: Filtering by labels is not currently supported")
+
+		output, err := start(trace.Spec.Gadget, trace.Spec.Node, filteredContainers)
+		if err != nil {
+			gadgets.SetStatusError(trace, err.Error())
 			return
 		}
-		if trace.Spec.Filter.ContainerName != "" {
-			log.Warningf("Gadget %s: Container name filter is not applicable in this gadget, ignoring it!", trace.Spec.Gadget)
-		}
 
-		pidsInPod := t.resolver.LookupPIDByPod(trace.Spec.Filter.Namespace, trace.Spec.Filter.Podname)
+		trace.Status.OperationError = ""
+		trace.Status.Output = output
+		trace.Status.State = "Completed"
+	default:
+		gadgets.SetStatusError(trace, fmt.Sprintf("Unknown operation %q", operation))
+	}
+}
+
+func start(gadgetName, node string, filteredContainers []pb.ContainerDefinition) (string, error) {
+	allSockets := []socketcollectortypes.Event{}
+
+	// Given that the socket-collector tracer works per network namespace and
+	// all the containers inside a pod/namespace share the network namespace,
+	// we only need to run the tracer with one valid PID per pod/namespace
+	visitedPods := make(map[string]bool)
+
+	// Keep also track of the number of different pods and the discarded
+	// containers to then check that we did not miss a pod
+	totalPods := 0
+	discardedContainers := []pb.ContainerDefinition{}
+
+	for _, container := range filteredContainers {
+		key := container.Podname + "/" + container.Namespace
+		if _, ok := visitedPods[key]; !ok {
+			totalPods++
+
+			if container.Pid == 0 {
+				discardedContainers = append(discardedContainers, container)
+
+				log.Warningf("Gadget %s: Discarding empty PID of Container %q Pod %q "+
+					"Namespace %q. Look for another Container in the same Pod",
+					gadgetName, container.Name, container.Podname, container.Namespace)
```

Comment:
Regarding my comment about containers with pid==0. I'd just print this warning message and that's all. I won't worry about counting containers and so on, I think it's too complicated and the benefit is very small.

Reply:
See my comment [above](https://github.com/kinvolk/inspektor-gadget/pull/247#discussion_r700336364) 😄